### PR TITLE
Support List<T> input and output

### PR DESCRIPTION
**Pull request by schneems**
Functions can now accept a list of strings `List<String>` from json:

```
$ ~/Documents/projects/work/sf-fx-runtime-nodejs/invoke.sh localhost:8080"['hello', 'world']"
# no error
```

Functions can now return a list of strings `List<String>` to json. Example function:

```
public class ExampleFunction implements SalesforceFunction<FunctionInput, List<String>> {
 private static final Logger LOGGER = LoggerFactory.getLogger(ExampleFunction.class);

 @override
 public List<String> apply(InvocationEvent<FunctionInput> event, Context context)
   throws Exception {

  List<Account> accounts = new ArrayList<>();
  List<String> foo = new ArrayList<>();
  foo.add("Lol");

  return foo;
 }
```

Output:

```
$ ~/Documents/projects/work/sf-fx-runtime-nodejs/invoke.sh localhost:8080 "{'hello': 'world'}"
# ...
* Connection #0 to host localhost left intact
["Lol"]* Closing connection 0
```

In addition functions support arbitrary typed lists `List<T>` as well

- Add an interface to JsonLibrary#deserializeListAt
  - Add JacksonReflectionJsonLibrary#deserializeListAt [Tested]
  - Add GsonReflectionJsonLibrary#deserializeListAt [Tested]
  - Add GsonJsonLibrary#deserializeListAt [Tested]
- Add ListParameterizedType class based on this comment https://stackoverflow.com/a/25223817/147390
- Detect List input and type of List<T> in SalesforceFunctionsProjectFunctionsScanner to use ListPayloadUnmarshaller
- Detect List input and type of List<T> in SalesforceFunctionsProjectFunctionsScanner to use ListFunctionResultMarshaller
- Update ListPayloadUnmarshaller to use jsonLibrary#deserializeListAt interface [Tested]
- Add test for `List<POJO>` and `List<Integer>` for ListFunctionResultMarshaller

### CODE REVIEW

**Comment by Malax** [comment]
Hey @schneems I took this and finished it so we can merge it before the week end. I tried to keep the commits as clean a possible to make it easy to follow what I did. Most of my commits have a description too. I almost refactored everything, streamlined it and added unit and integration tests for the new functionality that were missing before.

@codecov-commenter hadn't updated the comment yet when I was writing this, but the coverage actually improved after I added the tests, despite what the comment says.

I'll approve this now, please do the same and give this comment a thumbs up if you think it's good to go now (since you can't approve your own PRs). I asked @edmorley to give this a review as well.

**Review by Malax — APPROVED** [review]

File: sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/JsonLibraryDetector.java [hunk 1]
```diff
@@ -33,6 +34,22 @@ public static JsonLibrary detect(Class<?> clazz) throws AmbiguousJsonLibraryExce
...
+    return detect(clazz, classLoader);
+  }
+
+  /**
+   * Detects which JsonLibrary should be used with the given type. If no specific JsonLibrary is
+   * strictly required for the given type, a fallback JsonLibrary will be used as a catch-all.
+   *
+   * @see JsonLibrary#mustBeUsedFor(Type)
+   * @param type The type to detect the JSON library for.
+   * @param classLoader The class loader to use to load classes for given type.
```
Comment (edmorley) [inline sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/JsonLibraryDetector.java:46]:
```suggestion
   * @param classLoader The class loader to use to load classes for the given type.
```

File: sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/ListParameterizedType.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+package com.salesforce.functions.jvm.runtime.json;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.List;
+
+public class ListParameterizedType implements ParameterizedType {
```
Comment (edmorley) [inline sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/ListParameterizedType.java:13]: Does this need a javadoc definition? I don't know what the convention is for when to include them.

**Review by edmorley — APPROVED** [review]

File: sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/ListParameterizedType.java (hunk 2 above)
Comment (schneems) [reply sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/ListParameterizedType.java:31]: I took a stab at docs. It looks like most of the javadocs are method docs. I personally prefer a top level class doc. We can talk specifics when Manuel is back.

Comment (Malax) [reply sf-fx-runtime-java-runtime/src/main/java/com/salesforce/functions/jvm/runtime/json/ListParameterizedType.java:31]: Thanks for taking a stab at docs @schneems! 

In this case, I don't think we really need them here. This class implements a [well documented](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/ParameterizedType.html) interface of the standard library. The usage of the interface itself is documented there and this class does not do anything fancy and just returns fixed values in each method implementation.

If anyone comes across this class and wonders what it does and what it's for, they can read the `ParameterizedType` docs which are usually just a click away.

Side-note: Standard Javadoc does not support Markdown.

**Comment by edmorley** [comment]
@schneems CI didn't trigger - maybe you need to "follow" the Circle CI project? I've triggered manually via the branch view:
https://app.circleci.com/pipelines/github/forcedotcom/sf-fx-runtime-java?branch=schneems%2Fsupport-listT-input-output

**Comment by edmorley** [comment]
I've rebased this PR and removed the last commit updating the docs (since the docs linter was failing and it sounds like we don't need those changes: https://github.com/forcedotcom/sf-fx-runtime-java/pull/126#discussion_r689341572).

CI is now passing, so I think this is good to merge - @Malax do you agree? :-)

**Pull request merged**